### PR TITLE
[IMP] Improve and simplified customer form

### PR DESCRIPTION
**Pull request by gan-odoo**
Task : https://www.odoo.com/web#id=13765&view_type=form&model=project.task&action=327
Pad : https://pad.odoo.com/p/r.rwwROfDJuoEdD3by

### CODE REVIEW

File: addons/account/models/partner.py [hunk 1]
```diff
@@ -408,3 +408,11 @@ def _commercial_fields(self):
         return super(ResPartner, self)._commercial_fields() + \
             ['debit_limit', 'property_account_payable_id', 'property_account_receivable_id', 'property_account_position_id',
              'property_payment_term_id', 'property_supplier_payment_term_id', 'last_time_entries_checked']
+
+    @api.multi
+    def action_open_partner_bank(self):
```
Comment (atp-odoo) [inline addons/account/models/partner.py]: use ensure_one().

File: addons/account/models/partner.py [hunk 2]
```diff
@@ -408,3 +408,11 @@ def _commercial_fields(self):
...
+        result = self.env.ref('base.action_res_partner_bank_account_form').read()[0]
+        if len(self.bank_ids) <= 1:
+            result['res_id'] = self.bank_ids and self.bank_ids.id or False
```
Comment (atp-odoo) [inline addons/account/models/partner.py]: no need to check self.bank_ids. it's recordset.

File: openerp/addons/base/res/res_partner.py [hunk 3]
```diff
@@ -548,12 +548,17 @@ def open_commercial_entity(self, cr, uid, ids, context=None):
     def open_parent(self, cr, uid, ids, context=None):
         """ Utility method used to add an "Open Parent" button in partner views """
         partner = self.browse(cr, uid, ids[0], context=context)
-        return {'type': 'ir.actions.act_window',
-                'res_model': 'res.partner',
-                'view_mode': 'form',
-                'res_id': partner.parent_id.id,
-                'target': 'new',
-                'flags': {'form': {'action_buttons': True}}}
+        address_view_id = self.pool.get('ir.model.data').xmlid_to_res_id(cr, uid, 'base.view_partner_address_form')
```
Comment (atp-odoo) [inline openerp/addons/base/res/res_partner.py]: use self.pool.

File: openerp/addons/base/res/res_partner.py [hunk 4]
```diff
@@ -548,12 +548,17 @@ def open_commercial_entity(self, cr, uid, ids, context=None):
...
+        return {
+            'name': _('Customer'),
```
Comment (atp-odoo) [inline openerp/addons/base/res/res_partner.py]: what if we open supplier address form . ?

File: openerp/addons/base/res/res_partner_view.xml [hunk 5]
```diff
@@ -121,6 +121,61 @@
...
+                            <field name="title" options='{"no_open": True}'/>
+                            <field name="lang"/>
+                        </group>
+                    </group>
+                </form>
+            </field>
+        </record>
+
+        <record id="action_partner_customer_address_form" model="ir.actions.act_window">
+            <field name="name">Customers</field>
```
Comment (atp-odoo) [inline openerp/addons/base/res/res_partner_view.xml]: no need of this action. only view is required.

**Comment by gan-odoo** [comment]
@atp-odoo  I have all changes as per your comment on PR.
Thanks.

File: addons/base_vat/base_vat.py [hunk 6]
```diff
@@ -119,6 +119,17 @@ def check_vat(self, cr, uid, ids, context=None):
...
+        if vals.get('vat'):
+            self.validate_vat(cr, uid, ids, context=context)
+        return super(res_partner, self).write(cr, uid, ids, vals, context=context)
+
+    def create(self, cr, uid, vals, context=None):
+        new_id = super(res_partner, self).create(cr, uid, vals, context=context)
+        if vals.get('vat'):
+            self.validate_vat(cr, uid, [new_id], context=context)
+        return new_id
+
```
Comment (atp-odoo) [inline addons/base_vat/base_vat.py]: no need of create and write with  validate_vat method if button is removed. constraints can check vat and validate it.

File: addons/account/models/partner.py [hunk 7]
```diff
@@ -285,6 +285,11 @@ def _journal_item_count(self):
             partner.journal_item_count = self.env['account.move.line'].search_count([('partner_id', '=', partner.id)])
             partner.contracts_count = self.env['account.analytic.account'].search_count([('partner_id', '=', partner.id)])
 
+    @api.multi
```
Comment (antonylesuisse) [inline addons/account/models/partner.py]: for stat button fields try whenever possible to be multi and to do the job in O(1) sql queries, using count or group by.

File: addons/payment/models/res_partner.py [hunk 8]
```diff
@@ -4,6 +4,14 @@
 class res_partner(osv.osv):
     _name = 'res.partner'
     _inherit = 'res.partner'
+
+    def _get_payment_method_count(self, cr, uid, ids, field_names, arg, context=None):
```
Comment (antonylesuisse) [inline addons/payment/models/res_partner.py]: same same

File: addons/account/models/partner.py (hunk 7 above)
Comment (gan-odoo) [inline addons/account/models/partner.py]: @antonylesuisse Here also use api.multi. Why we use read_group or sql query here. When bank_ids is one2many field that count bank_ids is easily get using len().
is it required to use read_group here ?

Comment (antonylesuisse) [inline addons/account/models/partner.py]: Is it possible to do it with read_group or not ?

Comment (atp-odoo) [inline addons/account/models/partner.py]: yes it's possible. and we will do it.
Thanks @antonylesuisse

Comment (gan-odoo) [inline addons/account/models/partner.py]: Done!

File: addons/payment/models/res_partner.py (hunk 8 above)
Comment (gan-odoo) [inline addons/payment/models/res_partner.py]: Done

File: addons/crm/res_partner.py [hunk 9]
```diff
@@ -12,7 +12,7 @@ def _opportunity_meeting_count(self, cr, uid, ids, field_name, arg, context=None
         # the user may not have access rights for opportunities or meetings
         try:
             for partner in self.browse(cr, uid, ids, context):
-                if partner.is_company:
+                if partner.is_company == 'company':
```
Comment (tivisse) [inline addons/crm/res_partner.py]: A boolean becomes a string ? Well, that's unexpected...

Comment (gan-odoo) [inline addons/crm/res_partner.py]: Now 'is_company' is a selection field not boolean. (See this commit : https://github.com/odoo/odoo/commit/67447b2fc9eb5436e7ac2b5098373cd0f06b7860)

Comment (tivisse) [inline addons/crm/res_partner.py]: Then change the field name please.

Comment (gan-odoo) [inline addons/crm/res_partner.py]: @tivisse  changed the field name 'is_company' replace with 'company_type'.

File: addons/crm/base_partner_merge.py [hunk 10]
```diff
@@ -643,7 +643,7 @@ def parent_migration_process_cb(self, cr, uid, ids, context=None):
             UPDATE
                 res_partner
             SET
-                is_company = NULL,
+                company_type = NULL,
```
Comment (atp-odoo) [inline addons/crm/base_partner_merge.py]: company_type selection field can be Null ?

File: addons/crm/base_partner_merge.py [hunk 11]
```diff
@@ -694,10 +694,10 @@ def update_all_process_cb(self, cr, uid, ids, context=None):
             UPDATE
                 res_partner
             SET
-                is_company = NULL
+                company_type = NULL
```
Comment (atp-odoo) [inline addons/crm/base_partner_merge.py]: same here can we set company_type Null ?

File: openerp/addons/base/res/res_partner.py [hunk 12]
```diff
@@ -339,16 +322,11 @@ def copy(self, default=None):
         return super(res_partner, self).copy(default)
 
     @api.multi
-    def onchange_type(self, is_company):
+    def onchange_type(self, company_type):
         value = {'title': False}
```
Comment (atp-odoo) [inline openerp/addons/base/res/res_partner.py]: no more required this method. both field 'use_parent_address' and 'domain' removed.

Comment (gan-odoo) [inline openerp/addons/base/res/res_partner.py]: Yes, No need this method.

File: addons/crm/base_partner_merge.py (hunk 10 above)
Comment (gan-odoo) [inline addons/crm/base_partner_merge.py]: company_type is 'person' replaced with Null.

File: addons/crm/base_partner_merge.py (hunk 11 above)
Comment (gan-odoo) [inline addons/crm/base_partner_merge.py]: company_type is 'person' replaced with Null.

File: addons/website_customer/views/website_customer_view.xml [hunk 13]
```diff
@@ -1,20 +1,6 @@
...
     <data>
-        <record id="view_partners_form_website" model="ir.ui.view">
-            <field name="name">view.res.partner.form.website.tags</field>
-            <field name="model">res.partner</field>
-            <field name="inherit_id" ref="website_partner.view_partners_form_website" />
-            <field name="arch" type="xml">
-                <data>
-                    <xpath expr="//group[@name='mailing']" position="after">
-                        <group name="website" string="Website">
-                            <field name="tag_ids" widget="many2many_tags"/>
```
Comment (tde-banana-odoo) [inline addons/website_customer/views/website_customer_view.xml]: Ah tags are not editable from the front-end, you cannot remove it here.

Comment (gan-odoo) [inline addons/website_customer/views/website_customer_view.xml]: @tde-banana-odoo We are created already new task for add tags from front-end.
Task : https://www.odoo.com/web#id=16142&view_type=form&model=project.task&action=333&active_id=131

**Comment by tde-banana-odoo** [comment]
Clean commits branch: https://github.com/odoo/odoo/pull/8045 .

**Pull request closed without merging**